source/common/config/delta_subscription_state.cc

```diff
@@ -14,37 +14,80 @@ namespace Config {
+      type_url_(std::move(type_url)), watch_map_(watch_map), local_info_(local_info),
+      dispatcher_(dispatcher) {}
 
 void DeltaSubscriptionState::updateSubscriptionInterest(
     const absl::flat_hash_set<std::string>& cur_added,
     const absl::flat_hash_set<std::string>& cur_removed) {
   for (const auto& a : cur_added) {
-    setResourceWaitingForServer(a);
+    if (in_initial_legacy_wildcard_ && a != Wildcard) {
+      in_initial_legacy_wildcard_ = false;
+    }
+    // If the requested resource existed as a wildcard resource,
+    // transition it to requested. Otherwise mark it as a resource
+    // waiting for the server to receive the version.
+    if (auto it = wildcard_resource_state_.find(a); it != wildcard_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      wildcard_resource_state_.erase(it);
+    } else if (it = ambiguous_resource_state_.find(a); it != ambiguous_resource_state_.end()) {
+      requested_resource_state_.insert_or_assign(a, ResourceState::withVersion(it->second));
+      ambiguous_resource_state_.erase(it);
+    } else {
+      requested_resource_state_.insert_or_assign(a, ResourceState::waitingForServer());
+    }
+    ASSERT(requested_resource_state_.contains(a));
+    ASSERT(!wildcard_resource_state_.contains(a));
+    ASSERT(!ambiguous_resource_state_.contains(a));
     // If interest in a resource is removed-then-added (all before a discovery request
     // can be sent), we must treat it as a "new" addition: our user may have forgotten its
     // copy of the resource after instructing us to remove it, and need to be reminded of it.
     names_removed_.erase(a);
     names_added_.insert(a);
   }
   for (const auto& r : cur_removed) {
-    removeResourceState(r);
+    in_initial_legacy_wildcard_ = false;
+    // The resource we are interested in could also come from a wildcard subscription. Instead of
+    // removing it outright, mark the resource as not interesting to us any more. The server could
+    // later send us an update. If we don't have a wildcard subscription, just drop it.
+    if (requested_resource_state_.contains(Wildcard)) {
+      if (auto it = requested_resource_state_.find(r); it != requested_resource_state_.end()) {
+        // Wildcard resources always have a version. If our requested resource has no version, it
+        // won't be a wildcard resource then. If r is Wildcard itself, then it never has a version
+        // attached to it.
+        if (!it->second.isWaitingForServer()) {
+          ambiguous_resource_state_.insert({it->first, it->second.version()});
+        }
+        requested_resource_state_.erase(it);
+      }
+    } else {
+      requested_resource_state_.erase(r);
+    }
+    ASSERT(!requested_resource_state_.contains(r));
+    // This function shouldn't ever be called for resources that came from wildcard subscription.
+    ASSERT(!wildcard_resource_state_.contains(r));
```

Comment:
I find this confusing: there's a check above for a case we might be handling a resource belonging to a wildcard subscription, but here we state that is should never be the case.

Reply:
> I find this confusing: there's a check above for a case we might be handling a resource initially belonging to a wildcard subscription, but here we check that it isn't currently a part of a wildcard subscription. Can't we just drop it altogether -- the server will resend the resource if it still belongs to the wildcard subscription anyway?
> 

What we check above is whether we are subscribed to the wildcard resource. If so, we move the removed resource (not wildcard) from requested to ambiguous, because at this point we don't know if the removed resource is also a part of the wildcard resources set. So we try to be conservative and keep the resource in our cache until the server clarifies the situation. The assertions basically ensure that at all times a resource in cache is exactly in only one of three categories (requested, wildcard or ambiguous).

> Also: this assert should probably be done at the beginning of the method, rather then at the end -- it makes it clear what the constraints are.

You probably meant in the beginning of the loop body? I could do that, but I wanted to keep the asserts in one place and have them after the possible modifications of the cache.

Reply:
> So we try to be conservative and keep the resource in our cache until the server clarifies the situation.

I don't understand the purpose caching the resource: if the resource does belong to a wildcard subscription, the server will resend it in full anyway. Another issue is if the resource is standalone: how does it get removed from the ambiguous state?

Reply:
> You probably meant in the beginning of the loop body? I could do that, but I wanted to keep the asserts in one place and have them after the possible modifications of the cache.

I think it's better to keep them at the beginning of the method -- they are easier to spot this way and they do reveal important constraints.

Reply:
> > So we try to be conservative and keep the resource in our cache until the server clarifies the situation.
> 
> I don't understand the purpose of caching the resource: if the resource does belong to a wildcard subscription, the server will resend it in full anyway. Another issue is if the resource is standalone: how does it get removed from the ambiguous state?

I think this is being addressed by #17983.

> I think it's better to keep them at the beginning of the method -- they are easier to spot this way and they do reveal important constraints.

These asserts are here mostly to verify that after modifying the cache, the invariant "resource in only one category at a time" is still held. That's why I added them after modifying the cache, not before. Otherwise I'd need to write a code like:

```c++
for (auto const& name : cur_removed) {
  ASSERT(!wildcard_resource_state_.contains(name));
}
```

The comment that states that the function should not be called to remove an interest in the resource from the wildcard set was just an observation I made. But I think I'll drop this assert anyway, since we probably don't want to crash when someone actually calls this function with such a resource.

Reply:
Assert removed.